File: reference/python-engine/color_engine/conversions.py
# Garau Transparency Studio - Color Conversions
"""
Color space conversion functions.

Optimized for M1 Ultra using numpy vectorization where possible.
Uses perceptually uniform spaces (LCH/Lab) for Garau's visual mixture logic.
"""

import numpy as np
from typing import Tuple
from .color_types import RGB, HSL, LCH, NormalizedRGB


def rgb_to_normalized(rgb: RGB) -> NormalizedRGB:
    """Convert 0-255 RGB to 0-1 normalized RGB."""
    return NormalizedRGB(
        r=rgb.r / 255.0,
        g=rgb.g / 255.0,
        b=rgb.b / 255.0
    )


def normalized_to_rgb(normalized: NormalizedRGB) -> RGB:
    """Convert 0-1 normalized RGB to 0-255 RGB."""
    return RGB(
        r=round(normalized.r * 255),
        g=round(normalized.g * 255),
        b=round(normalized.b * 255)
    )


def rgb_to_hsl(rgb: RGB) -> HSL:
    """
    Convert RGB to HSL color space.
    
    Args:
        rgb: RGB color with 0-255 values
        
    Returns:
        HSL with h: 0-360, s: 0-100, l: 0-100
    """
    r, g, b = rgb.r / 255.0, rgb.g / 255.0, rgb.b / 255.0
    
    max_c = max(r, g, b)
    min_c = min(r, g, b)
    l = (max_c + min_c) / 2.0
    
    if max_c == min_c:
        h = s = 0.0
    else:
        d = max_c - min_c
        s = d / (2.0 - max_c - min_c) if l > 0.5 else d / (max_c + min_c)
        
        if max_c == r:
            h = (g - b) / d + (6.0 if g < b else 0.0)
        elif max_c == g:
            h = (b - r) / d + 2.0
        else:
            h = (r - g) / d + 4.0
        
        h /= 6.0
    
    return HSL(h=h * 360.0, s=s * 100.0, l=l * 100.0)


def hsl_to_rgb(hsl: HSL) -> RGB:
    """
    Convert HSL to RGB color space.
    
    Args:
        hsl: HSL with h: 0-360, s: 0-100, l: 0-100
        
    Returns:
        RGB with 0-255 values
    """
    h, s, l = hsl.h / 360.0, hsl.s / 100.0, hsl.l / 100.0
    
    if s == 0:
        r = g = b = l
    else:
        def hue_to_rgb(p, q, t):
            if t < 0: t += 1
            if t > 1: t -= 1
            if t < 1/6: return p + (q - p) * 6 * t
            if t < 1/2: return q
            if t < 2/3: return p + (q - p) * (2/3 - t) * 6
            return p
        
        q = l * (1 + s) if l < 0.5 else l + s - l * s
        p = 2 * l - q
        r = hue_to_rgb(p, q, h + 1/3)
        g = hue_to_rgb(p, q, h)
        b = hue_to_rgb(p, q, h - 1/3)
    
    return RGB(
        r=round(r * 255),
        g=round(g * 255),
        b=round(b * 255)
    )


def rgb_to_lab(rgb: RGB) -> Tuple[float, float, float]:
    """
    Convert RGB to CIE Lab color space.
    Uses D65 illuminant and sRGB color space.
    
    Args:
        rgb: RGB color with 0-255 values
        
    Returns:
        Lab tuple (L: 0-100, a: -128 to 127, b: -128 to 127)
    """
    # First convert to linear RGB
    r, g, b = rgb.r / 255.0, rgb.g / 255.0, rgb.b / 255.0
    
    # Apply gamma correction (sRGB to linear)
    def gamma_correct(c):
        return c / 12.92 if c <= 0.04045 else ((c + 0.055) / 1.055) ** 2.4
    
    r, g, b = gamma_correct(r), gamma_correct(g), gamma_correct(b)
    
    # Convert to XYZ (D65 illuminant)
    x = r * 0.4124564 + g * 0.3575761 + b * 0.1804375
    y = r * 0.2126729 + g * 0.7151522 + b * 0.0721750
    z = r * 0.0193339 + g * 0.1191920 + b * 0.9503041
    
    # Reference white D65
    xn, yn, zn = 0.95047, 1.0, 1.08883
    
    def f(t):
        return t ** (1/3) if t > 0.008856 else (7.787 * t) + (16 / 116)
    
    L = (116 * f(y / yn)) - 16
    a = 500 * (f(x / xn) - f(y / yn))
    b_val = 200 * (f(y / yn) - f(z / zn))
    
    return (max(0, L), a, b_val)


def lab_to_rgb(lab: Tuple[float, float, float]) -> RGB:
    """
    Convert CIE Lab to RGB color space.
    
    Args:
        lab: Lab tuple (L: 0-100, a: -128 to 127, b: -128 to 127)
        
    Returns:
        RGB with 0-255 values
    """
    L, a, b_val = lab
    
    # Reference white D65
    yn = 1.0
    xn = 0.95047
    zn = 1.08883
    
    def f_inv(t):
        return t ** 3 if t > 0.20689303442296383 else (t - 16/116) / 7.787
    
    fy = (L + 16) / 116
    fx = fy + a / 500
    fz = fy - b_val / 200
    
    x = xn * f_inv(fx)
    y = yn * f_inv(fy)
    z = zn * f_inv(fz)
    
    # Convert to linear RGB
    r = x * 3.2404542 + y * -1.5371385 + z * -0.4985314
    g = x * -0.9692660 + y * 1.8760108 + z * 0.0415560
    b = x * 0.0556434 + y * -0.2040259 + z * 1.0572252
    
    # Apply gamma correction (linear to sRGB)
    def gamma_inverse(c):
        return 12.92 * c if c <= 0.0031308 else (1.055 * (c ** (1/2.4))) - 0.055
    
    r = max(0, min(1, gamma_inverse(r)))
    g = max(0, min(1, gamma_inverse(g)))
    b = max(0, min(1, gamma_inverse(b)))
    
    return RGB(
        r=round(r * 255),
        g=round(g * 255),
        b=round(b * 255)
    )


def rgb_to_lch(rgb: RGB) -> LCH:
    """
    Convert RGB to LCH (Lightness, Chroma, Hue) color space.
    LCH is perceptually uniform - crucial for Garau's visual mixture analysis.
    
    Args:
        rgb: RGB color with 0-255 values
        
    Returns:
        LCH with l: 0-100, c: 0-100+, h: 0-360
    """
    L, a, b_val = rgb_to_lab(rgb)
    
    # Convert from Cartesian (a, b) to polar (c, h)
    c = np.sqrt(a ** 2 + b_val ** 2)
    h = np.degrees(np.arctan2(b_val, a))
    h = h % 360  # Normalize to 0-360
    
    return LCH(l=L, c=c, h=h)


def lch_to_rgb(lch: LCH) -> RGB:
    """
    Convert LCH to RGB color space.
    
    Args:
        lch: LCH with l: 0-100, c: 0-100+, h: 0-360
        
    Returns:
        RGB with 0-255 values
    """
    # Convert from polar (c, h) to Cartesian (a, b)
    h_rad = np.radians(lch.h)
    a = lch.c * np.cos(h_rad)
    b_val = lch.c * np.sin(h_rad)
    
    return lab_to_rgb((lch.l, a, b_val))


def rgb_to_hex(rgb: RGB) -> str:
    """Convert RGB to hex string."""
    return rgb.to_hex()


def hex_to_rgb(hex_str: str) -> RGB:
    """Convert hex string to RGB."""
    return RGB.from_hex(hex_str)


def interpolate_rgb(c1: RGB, c2: RGB, t: float) -> RGB:
    """
    Interpolate between two RGB colors in perceptual space (LCH).
    
    Args:
        c1: Start color
        c2: End color
        t: Interpolation factor (0-1)
        
    Returns:
        Interpolated RGB color
    """
    lch1 = rgb_to_lch(c1)
    lch2 = rgb_to_lch(c2)
    
    # Interpolate in LCH space for perceptually uniform results
    l = lch1.l + t * (lch2.l - lch1.l)
    c = lch1.c + t * (lch2.c - lch1.c)
    
    # Handle hue interpolation (take shortest path)
    h1, h2 = lch1.h, lch2.h
    dh = h2 - h1
    if dh > 180:
        dh -= 360
    elif dh < -180:
        dh += 360
    h = (h1 + t * dh) % 360
    
    return lch_to_rgb(LCH(l=l, c=c, h=h))


def get_luminance(rgb: RGB) -> float:
    """
    Calculate relative luminance (perceived brightness).
    Uses ITU-R BT.709 coefficients.
    
    Args:
        rgb: RGB color
        
    Returns:
        Luminance value 0-1
    """
    r, g, b = rgb.r / 255.0, rgb.g / 255.0, rgb.b / 255.0
    
    # Apply gamma correction
    def linearize(c):
        return c / 12.92 if c <= 0.04045 else ((c + 0.055) / 1.055) ** 2.4
    
    r, g, b = linearize(r), linearize(g), linearize(b)
    
    # ITU-R BT.709 luminance coefficients
    return 0.2126 * r + 0.7152 * g + 0.0722 * b


def get_contrast_ratio(c1: RGB, c2: RGB) -> float:
    """
    Calculate WCAG contrast ratio between two colors.
    
    Args:
        c1: First color
        c2: Second color
        
    Returns:
        Contrast ratio (1:1 to 21:1)
    """
    L1 = get_luminance(c1)
    L2 = get_luminance(c2)
    
    lighter = max(L1, L2)
    darker = min(L1, L2)
    
    return (lighter + 0.05) / (darker + 0.05)

File: packages/engine/src/conversions.ts
/**
 * Color-space conversions.
 *
 * Uses perceptually uniform spaces (Lab / LCH) for Garau's visual-mixture
 * logic, and sRGB / ITU-R BT.709 luminance for Metelli's lightness conditions.
 * Ported from `reference/python-engine/color_engine/conversions.py`.
 */

import type { RGB, NormalizedRGB, HSL, Lab, LCH } from './types.js';

const clamp = (v: number, min = 0, max = 1): number => Math.max(min, Math.min(max, v));

/** Clamp + round an RGB triple into a valid 0–255 integer color. */
export function rgb(r: number, g: number, b: number): RGB {
  return {
    r: Math.round(clamp(r, 0, 255)),
    g: Math.round(clamp(g, 0, 255)),
    b: Math.round(clamp(b, 0, 255)),
  };
}

export function rgbToNormalized(c: RGB): NormalizedRGB {
  return { r: c.r / 255, g: c.g / 255, b: c.b / 255 };
}

export function normalizedToRgb(c: NormalizedRGB): RGB {
  return rgb(c.r * 255, c.g * 255, c.b * 255);
}

export function rgbToHsl(c: RGB): HSL {
  const r = c.r / 255;
  const g = c.g / 255;
  const b = c.b / 255;
  const maxC = Math.max(r, g, b);
  const minC = Math.min(r, g, b);
  const l = (maxC + minC) / 2;

  let h = 0;
  let s = 0;
  if (maxC !== minC) {
    const d = maxC - minC;
    s = l > 0.5 ? d / (2 - maxC - minC) : d / (maxC + minC);
    if (maxC === r) h = (g - b) / d + (g < b ? 6 : 0);
    else if (maxC === g) h = (b - r) / d + 2;
    else h = (r - g) / d + 4;
    h /= 6;
  }
  return { h: h * 360, s: s * 100, l: l * 100 };
}

export function hslToRgb(hsl: HSL): RGB {
  const h = hsl.h / 360;
  const s = hsl.s / 100;
  const l = hsl.l / 100;

  if (s === 0) return rgb(l * 255, l * 255, l * 255);

  const hueToRgb = (p: number, q: number, tIn: number): number => {
    let t = tIn;
    if (t < 0) t += 1;
    if (t > 1) t -= 1;
    if (t < 1 / 6) return p + (q - p) * 6 * t;
    if (t < 1 / 2) return q;
    if (t < 2 / 3) return p + (q - p) * (2 / 3 - t) * 6;
    return p;
  };

  const q = l < 0.5 ? l * (1 + s) : l + s - l * s;
  const p = 2 * l - q;
  return rgb(
    hueToRgb(p, q, h + 1 / 3) * 255,
    hueToRgb(p, q, h) * 255,
    hueToRgb(p, q, h - 1 / 3) * 255,
  );
}

const gammaToLinear = (c: number): number =>
  c <= 0.04045 ? c / 12.92 : ((c + 0.055) / 1.055) ** 2.4;

const linearToGamma = (c: number): number =>
  c <= 0.0031308 ? 12.92 * c : 1.055 * c ** (1 / 2.4) - 0.055;

// D65 reference white.
const Xn = 0.95047;
const Yn = 1.0;
const Zn = 1.08883;

export function rgbToLab(c: RGB): Lab {
  const r = gammaToLinear(c.r / 255);
  const g = gammaToLinear(c.g / 255);
  const b = gammaToLinear(c.b / 255);

  const x = r * 0.4124564 + g * 0.3575761 + b * 0.1804375;
  const y = r * 0.2126729 + g * 0.7151522 + b * 0.072175;
  const z = r * 0.0193339 + g * 0.119192 + b * 0.9503041;

  const f = (tv: number): number => (tv > 0.008856 ? Math.cbrt(tv) : 7.787 * tv + 16 / 116);
  const fy = f(y / Yn);
  return {
    l: Math.max(0, 116 * fy - 16),
    a: 500 * (f(x / Xn) - fy),
    b: 200 * (fy - f(z / Zn)),
  };
}

export function labToRgb(lab: Lab): RGB {
  const fInv = (tv: number): number => (tv > 0.20689303442296383 ? tv ** 3 : (tv - 16 / 116) / 7.787);
  const fy = (lab.l + 16) / 116;
  const fx = fy + lab.a / 500;
  const fz = fy - lab.b / 200;

  const x = Xn * fInv(fx);
  const y = Yn * fInv(fy);
  const z = Zn * fInv(fz);

  const r = x * 3.2404542 + y * -1.5371385 + z * -0.4985314;
  const g = x * -0.969266 + y * 1.8760108 + z * 0.041556;
  const b = x * 0.0556434 + y * -0.2040259 + z * 1.0572252;

  return rgb(linearToGamma(r) * 255, linearToGamma(g) * 255, linearToGamma(b) * 255);
}

export function rgbToLch(c: RGB): LCH {
  const { l, a, b } = rgbToLab(c);
  const chroma = Math.sqrt(a * a + b * b);
  let h = (Math.atan2(b, a) * 180) / Math.PI;
  h = ((h % 360) + 360) % 360;
  return { l, c: chroma, h };
}

export function lchToRgb(lch: LCH): RGB {
  const hRad = (lch.h * Math.PI) / 180;
  return labToRgb({ l: lch.l, a: lch.c * Math.cos(hRad), b: lch.c * Math.sin(hRad) });
}

export function rgbToHex(c: RGB): string {
  const h = (n: number): string => n.toString(16).padStart(2, '0').toUpperCase();
  return `#${h(c.r)}${h(c.g)}${h(c.b)}`;
}

export function hexToRgb(hex: string): RGB {
  const s = hex.replace(/^#/, '');
  const expanded = s.length === 3 ? s.split('').map((ch) => ch + ch).join('') : s;
  return rgb(
    parseInt(expanded.slice(0, 2), 16),
    parseInt(expanded.slice(2, 4), 16),
    parseInt(expanded.slice(4, 6), 16),
  );
}

/** Interpolate two RGB colors in perceptual LCH space (shortest hue path). */
export function interpolateRgb(c1: RGB, c2: RGB, t: number): RGB {
  const a = rgbToLch(c1);
  const b = rgbToLch(c2);
  const l = a.l + t * (b.l - a.l);
  const c = a.c + t * (b.c - a.c);
  let dh = b.h - a.h;
  if (dh > 180) dh -= 360;
  else if (dh < -180) dh += 360;
  const h = ((a.h + t * dh) % 360 + 360) % 360;
  return lchToRgb({ l, c, h });
}

/** Relative luminance (ITU-R BT.709), 0–1. */
export function getLuminance(c: RGB): number {
  const r = gammaToLinear(c.r / 255);
  const g = gammaToLinear(c.g / 255);
  const b = gammaToLinear(c.b / 255);
  return 0.2126 * r + 0.7152 * g + 0.0722 * b;
}

/** WCAG contrast ratio between two colors (1:1 to 21:1). */
export function getContrastRatio(c1: RGB, c2: RGB): number {
  const l1 = getLuminance(c1);
  const l2 = getLuminance(c2);
  const lighter = Math.max(l1, l2);
  const darker = Math.min(l1, l2);
  return (lighter + 0.05) / (darker + 0.05);
}
